lk2nd/smp/cpu-boot.c

```diff
@@ -23,6 +23,80 @@
 #define APC_PWR_GATE_CTL_GHDS_EN	BIT(0)
 #define APC_PWR_GATE_CTL_GHDS_CNT(cnt)	((cnt) << 24)
 
+/*
+ * Each CPU within a cluster has a qcom,kpss-acc-v2.
+ * There is just once L2 cache controller per cluster.
+ *
+ * APCS_KPSS_GLB_BASE takes the base of given qcom,kpss-acc-v2 and offsets
+ * back the single APCS_KPSS_GLB register which is the base for the L2
+ * cache controller in the cluster.
+ *
+ * Hence qcom,kpss-acc-v2@0b088000
+ *       qcom,kpss-acc-v2@0b0a8000
+ *       qcom,kpss-acc-v2@0b0b8000
+ *       qcom,kpss-acc-v2@0b098000
+ *       => APCS_KPSS_GLB_BASE@0b011000
+ *
+ *       qcom,kpss-acc-v2@0b188000
+ *       qcom,kpss-acc-v2@0b198000
+ *       qcom,kpss-acc-v2@0b1a8000
+ *       qcom,kpss-acc-v2@0b1b8000
+ *       => APCS_KPSS_GLB_BASE@0b110000
+ */
+#define APCS_KPSS_GLB_MASK			0xfff00000
+#define APCS_KPSS_GLB_OFFSET			0x00011000
+#define APCS_KPSS_GLB_BASE(base)		(((base) & APCS_KPSS_GLB_MASK) + \
+						  APCS_KPSS_GLB_OFFSET)
+
+#define APCS_PWR_CTL_OVERRIDE_ADDR		0xc
+#define APCS_PWR_CTL_OVR_PRESETDBG		BIT(22)
+#define APCS_PWR_CTL_OVR_L1_RAM_COLLAPSE_DIS	(BIT(21) | BIT(20) | BIT(19) | BIT(18))
+#define APCS_PWR_CTL_OVR_SCU_RAM_COLLAPSE_DIS	BIT(17)
+#define APCS_PWR_CTL_OVR_L2_RET_SLP_DIS		BIT(16)
+#define APCS_PWR_CTL_OVR_CLK_EN_L2AUX_C		(BIT(2) | BIT(0))
+#define APCS_PWR_CTL_OVR_CLK_EN_L2PLL_C		BIT(2)
+#define APCS_PWR_CTL_OVR_CLK_EN_PISLEEP_CLK	(BIT(0) | BIT(1))
+#define APCS_PWR_CTL_OVR_CLK_EN_REF_CLK		BIT(1)
+#define APCS_PWR_CTL_OVR_CLK_EN_AHB_CLK		BIT(0)
+#define APCS_PWR_CTL_OVR_CLK_EN_QSB_CLK		0
+
+#define APCS_L2_PWR_CTL_ADDR			0x14
+#define APCS_PMIC_APC_O				BIT(28)
+#define APCS_L2_HS_CNT(cnt)			((cnt) << 16)
+#define APCS_L2_SCU_MAGIC_BIT1			BIT(15)
+#define APCS_L2_SCU_MAGIC_BIT0			BIT(14)
+#define APCS_L2_PWR_L2_RET_SLP			BIT(13)
+#define APCS_L2_PWR_CTL_SYS_RESET		BIT(12)
+#define APCS_L2_PWR_CTL_L2_SLEEP_STATE		BIT(11)
+#define APCS_L2_PWR_CTL_L2_HS_RST		BIT(10)
+#define APCS_L2_PWR_CTL_L2_HS_EN		BIT(9)
+#define APCS_L2_PWR_CTL_L2_HS_CLAMP		BIT(8)
+#define APCS_L2_PWR_CTL_L2_ARRAY_HS_3		BIT(3)
+#define APCS_L2_PWR_CTL_L2_ARRAY_HS_2		BIT(2)
+#define APCS_L2_PWR_CTL_L2_ARRAY_HS_1		BIT(1)
+#define APCS_L2_PWR_CTL_L2_ARRAY_HS_0		BIT(0)
```

Comment:
Hmm, great the TRM and HRD seems to have some disagreement about the naming here. But it looks like the TRM is more correct, so this should be:

```c
#define APCS_L2_PWR_CTL_L2_RST_DIS		BIT(2)
#define APCS_L2_PWR_CTL_SCU_ARRAY_HS		BIT(1)
#define APCS_L2_PWR_CTL_L2_ARRAY_HS		BIT(0)
```

and bit 3 likely doesn't exist in the hardware...

If you have contacts at Qualcomm, feel free to thank them for creating such confusing documentation. :) (Although I'm honestly really happy Qualcomm decided to have any public documentation at all for APQ8016E...)